autopilot/contractor.go

```diff
@@ -466,4 +487,75 @@ func (c *contractor) renewFundingEstimate(cfg Config, s State, cID types.FileCon
 	return estimatedCost, nil
 }
 
-func logErr(err error) {} // TODO
+func (c *contractor) candidateHosts(cfg Config, wanted int) ([]consensus.PublicKey, error) {
+	// fetch all candidate hosts
+	hosts, err := c.ap.bus.CandidateHosts()
+	if err != nil {
+		return nil, err
+	}
+
+	// apply host filter
+	filterFn := hostFilter(
+		cfg.isWhiteListed,
+		cfg.isBlackListed,
+		// TODO: isGouging,
+		isOffline,
+	)
+	hosts = hosts[:0]
+	for _, h := range hosts {
+		if result := filterFn(host{h, rhpv2.Contract{}}); !result.filtered() {
+			hosts = append(hosts, h)
+		}
+	}
+
+	// update num wanted
+	if wanted > len(hosts) {
+		wanted = len(hosts)
+	}
+
+	// score each host
+	scoreFn := HostScore(
```

Comment:
I don't see the advantage of passing around lambda functions for the host score versus just having a host score method that calls sub-methods and sums up the results.